### ISSUE TITLE
In-memory, synchronous database

### ISSUE THREAD

**Issue by puppybits** [prescottprue/redux-firestore#323]
**What is the feature?**
When doing something like drag and drop the updates to firestore take too long so the UI jumps. This is even worse with transactions.

**What version would this apply to?**
v0.15

**Do you have thoughts/examples on how the feature would be used (i.e. API)?**

Yes. I already forked it and made it work: https://github.com/TARAAI/redux-firestore/pull/1

```js
it('override a document add synchronously', () => {
      const doc1 = { key1: 'value1', id: 'testDocId1', path }; 
      const doc2 = { key1: 'value1', id: 'testDocId2', path }; 

      // Initial seed
      const action1 = {
        meta: {
          collection,
          storeAs: 'testStoreAs',
          where: [['key1', '==', 'value1']],
        },
        payload: { data: { [doc1.id]: doc1 }, ordered: [doc1] },
        type: actionTypes.LISTENER_RESPONSE,
      };
      const action2 = {
        type: actionTypes.OPTIMISTIC_ADDED,
        meta: {
          collection,
          doc: doc2.id,
        },
        payload: { data: { [doc2.id]: doc2 } },
      };

      const pass1 = reducer(initialState, action1);
      const pass2 = reducer(pass1, action2);
      
      expect(pass1.optimistic.testStoreAs.results[0]).to.eql(doc1);
      expect(pass1.optimistic.testStoreAs.results[1]).to.eql(undefined);
      expect(pass2.optimistic.testStoreAs.results[0]).to.eql(doc1);
      expect(pass2.optimistic.testStoreAs.results[1]).to.eql(doc2);
    });
```

**Comment by prescottprue** [comment]
Interesting approach, thanks for posting! I would say one downside to this methodology is that if the write breaks a database rule (not confirmed until after the promise completes), then we would also need a way to revert the optimistic change. We would probably also want this to be opt in since it could cause a flash for some cases (would not be expected)

Have you run into those cases?

**Comment by puppybits** [comment]
@prescottprue Yes there is an action to explicitly remove the optimistic commit. It removes automatically on write but failures/rollbacks aren’t handled yet. 

It needs some hardening (which is why I didn’t push a PR) to handle 100% of cases. The plan is that when there’s a failure it would need to have a unique action id to tie a batch, transaction or commit to a group of optimitisic updates.

**Comment by puppybits** [comment]
@prescottprue I've updated our fork and integrated it into our app. Here's an illustration of the issue and the difference between the 2 different reducers. 

**Before.** 
![ezgif-2-b71ab9e2110f](https://user-images.githubusercontent.com/140163/109244957-7e22b100-7794-11eb-9129-01fe0ed9ac1a.gif)

**After.**
![ezgif-2-acf3f960c486](https://user-images.githubusercontent.com/140163/109244971-837ffb80-7794-11eb-999a-ed8b757b05ba.gif)

We're doing internal QA on our app but when it all passes we'd be happy to send a PR. It does handle both success and failure cases when the firestore save reaches a resolution. Here's a link to our internal PR review to see the details https://github.com/TARAAI/redux-firestore/pull/4. 

Also we have an additional follow-on feature planned which would unify single document changes, batches, transactions and optimistic updates into a single API call so manually sending an optimistic update would no longer be needed.

**Comment by prescottprue** [comment]
@puppybits awesome! Thanks for the work making the reducer and sharing - looking forward to the PR

Really like the idea of optimistic updates, so let me know if there is anything I can do to speed up the process of getting this in

**Comment by puppybits** [comment]
We've tested the cache reducer locally and it's all working. But the optimistic update message API was only designed to be used internally. It's too verbose for a consumer. 

We just pushed an internal PR to the fork. It both unifies updating the cache reducer and firestore solo, batching and transactions into and single API. It's going through PR review and testing this week. 

https://github.com/TARAAI/redux-firestore/pull/5

### Changes included:

1. Simpler API for changing firestore documents.
2. Automatically updated the cache reducer to instantly reflect changes in Redux
3. Exposed a config to optionally enable onSnapshot includeMetadataChange
4. Added a config for a global FirestoreDataConvertor (We generate global decoder from our Typescript definitions).

```js
const transactionPromise = firestore.mutate({
  read: { 
     myLockedDocument:  { 
        collection: 'path', 
        doc: 'document-id',
     }
  },
  write: [({ myLockedDocument }) => ({
      collection: 'path', 
      doc: 'document-id' , 
      someNumber: myLockedDocument.someNumber + 10
    })]
});

// cache reducer has already added the requested changes to 
// the in-memory overrides as an optimistic update

transactionPromise.then(() => {
  /* change has completed from firestore*/
  /* cache reducer has already removed the optimistic update and is using firestore's full query results */
}).catch(() => {
  /* failures will also trigger the cache to remove the optimistic overrides */
  /* use any of the existing failure messages to be notified and create custom failure functionality */
});
```

**Comment by puppybits** [comment]
@prescottprue Submitted a PR for this new reducer. The headline is that is' 100x faster for all changes. 

The reason is that even single save operations have to round-trip over the network before the snapshot listener fires. For us saving a Firestore transaction in production took a minimum of 1.2s to show the results to the user. That dropped to 6ms. That's over 100x faster.

**Comment by cabello** [comment]
I am suffering from this issue also as I implemented drag & drop in one of by `useFirestoreConnect` queries.

@puppybits super excited about taking a look at your PR https://github.com/prescottprue/redux-firestore/pull/335 I noticed the changeset is 68k lines @prescottprue do you have a suggestion on breaking this down into something that could be merged into main and released?

I agree with the architectural, data flow issue: the React component is backed by Redux connected to Firebase, the drag & drop sends the change request to Firebase, that takes ~150ms (most times is faster), the UI flickers between old state and new state, would be lovely to have an optimistic UI system in place for these. ❤️
